Tests/TuistKitTests/Generator/TuistGeneratorModelLoaderTest.swift

```diff
@@ -0,0 +1,314 @@
+        // Given
+        try fileHandler.touch(path.appending(component: "model.xcdatamodeld"))
+        let manifest = ProjectDescription.CoreDataModel("model.xcdatamodeld",
+                                                        currentVersion: "1")
+        
+        // When
+        let model = try TuistKit.CoreDataModel.from(json: manifest.toJSON(), path: path, fileHandler: fileHandler)
+        
+        // Then
+        XCTAssertTrue(coreDataModel(model, matches: manifest, at: path))
+    }
+    
+    func test_targetActions() throws {
+        // Given
+        let manifest = ProjectDescription.TargetAction.test(name: "MyScript",
+                                                            tool: "my_tool",
+                                                            path: "my/path",
+                                                            order: .pre,
+                                                            arguments: ["arg1", "arg2"])
+        // When
+        let model = try TuistKit.TargetAction.from(json: manifest.toJSON(), path: path, fileHandler: fileHandler)
+        
+        // Then
+        XCTAssertEqual(model.name, "MyScript")
+        XCTAssertEqual(model.tool, "my_tool")
+        XCTAssertEqual(model.path, path.appending(RelativePath("my/path")))
+        XCTAssertEqual(model.order, .pre)
+        XCTAssertEqual(model.arguments, ["arg1", "arg2"])
+    }
+    
+    // MARK: - Helpers
+    
+    func createManifestLoader(with manifests: [AbsolutePath: Encodable]) -> GraphManifestLoading {
+        let manifestLoader = MockGraphManifestLoader()
+        manifestLoader.loadStub = { _, path in
+            guard let manifest = manifests[path] else {
+                throw GraphLoadingError.manifestNotFound(path)
+            }
+            return try manifest.toJSON()
+        }
+        return manifestLoader
+    }
+    
+    func assert(target: TuistKit.Target,
+                matches manifest: ProjectDescription.Target,
+                at path: AbsolutePath,
+                file: StaticString = #file,
+                line: UInt = #line) {
+        XCTAssertEqual(target.name, manifest.name, file: file, line: line)
+        XCTAssertEqual(target.bundleId, manifest.bundleId, file: file, line: line)
+        XCTAssertTrue(target.platform == manifest.platform, file: file, line: line)
+        XCTAssertTrue(target.product == manifest.product, file: file, line: line)
+        XCTAssertEqual(target.infoPlist, path.appending(RelativePath(manifest.infoPlist)), file: file, line: line)
+        XCTAssertEqual(target.entitlements, manifest.entitlements.map { path.appending(RelativePath($0)) }, file: file, line: line)
+        XCTAssertEqual(target.environment, manifest.environment, file: file, line: line)
+        assert(coreDataModels: target.coreDataModels, matches: manifest.coreDataModels, at: path, file: file, line: line)
+        optionalAssert(target.settings, manifest.settings, file: file, line: line) {
```

Comment:
This is quite verbose - will explore alternate options to tidy this up.

One idea I was playing with was using key paths to help with this. 

e.g.

```swift
let matchers: [Match<TuistKit.Target, ProjectDescription.Target>] = [
     .equal(\.name, \.name),
     .equal(\.bundleId, \.bundleId),
]
```

It ended up getting quite complex especially for nested structures, so opted to explore that separately and introduce it once it's ready. (anyone know of anything similar?)

Reply:
> It ended up getting quite complex especially for nested structures

Can you elaborate? We use something quite similar with keyPaths...

Reply:
I started out with a simple helper:

```swift
struct Matcher<Left, Right> {
    private let matcher: (Left, Right) -> Bool
    func evaluate(left: Left, right: Right) -> Bool {
        return matcher(left, right)
    }
}

extension Matcher {
    static func equal<Left, Right, Element: Equatable>(_ left: KeyPath<Left, Element>,
                                                       _ right: KeyPath<Right, Element>) -> Matcher<Left, Right> {
        return .equal(left, right) { $0 == $1 }
    }
    
    static func equal<Left, Right, LeftElement, RightElement>(_ left: KeyPath<Left, LeftElement>,
                                                       _ right: KeyPath<Right, RightElement>,
                                                       equate: @escaping (LeftElement, RightElement) -> Bool) -> Matcher<Left, Right> {
        return Matcher<Left, Right>(matcher: {
            let leftValue = $0[keyPath: left]
            let rightValue = $1[keyPath: right]
            return equate(leftValue, rightValue)
        })
    }
}
```

This would allow assertions:

```swift

let matcher[Matcher<TuistKit.Target, ProjectDescription.Target>] = [
     .equal(\.name, \.name),
     .equal(\.bundleId, \.bundleId),
    // ...
]

matchers.forEach {
   XCTAssertTrue($0.evaluate(target, manifest))
}
```

But then trying to get nested matchers to work (e.g. appending matchers for Settings) got quite tricky - also didn't like the fact I couldn't get a string representation of the KeyPath to help make it clearer what failed.

https://forums.swift.org/t/pitch-improving-keypath/6541

It's an interesting Swift exercise :) will play around some more with this concept outside of this PR see if it comes to anything (and research if something already exists like this) - suggestions / feedback welcome of course. :)

Reply:
Not too dissimilar but I was playing around building a predicate system for filter operations on sequences: https://github.com/ollieatkinson/Predicate.swift

The above implementation looks good. you could maybe even overload the `==` operator here to make it even more obvious it's an equality check.

e.g.

```swift
let matcher[Matcher<TuistKit.Target, ProjectDescription.Target>] = [
     \.name == \.name,
     \.bundleId == \.bundleId,
    // ...
]
```

Building some API around this could be really powerful. It is definitely worth exploring.